Source/Core/Core/PowerPC/CachedInterpreter/CachedInterpreter.cpp

```diff
@@ -177,94 +89,89 @@ void CachedInterpreter::SingleStep()
+s32 CachedInterpreter::HLEFunction(PowerPC::PowerPCState& ppc_state,
+                                   const HLEFunctionOperands& operands)
 {
-  auto& ppc_state = cached_interpreter.m_ppc_state;
-  ppc_state.pc = data.hex;
-  ppc_state.npc = data.hex + 4;
+  const auto& [system, current_pc, hook_index] = operands;
+  ppc_state.pc = current_pc;
+  HLE::Execute(Core::CPUThreadGuard{system}, current_pc, hook_index);
+  return sizeof(AnyCallback) + sizeof(operands);
 }
 
-void CachedInterpreter::WriteBrokenBlockNPC(CachedInterpreter& cached_interpreter,
-                                            UGeckoInstruction data)
+s32 CachedInterpreter::WriteBrokenBlockNPC(PowerPC::PowerPCState& ppc_state,
+                                           const WriteBrokenBlockNPCOperands& operands)
 {
-  cached_interpreter.m_ppc_state.npc = data.hex;
+  const auto& [current_pc] = operands;
+  ppc_state.npc = current_pc;
+  return sizeof(AnyCallback) + sizeof(operands);
 }
 
-bool CachedInterpreter::CheckFPU(CachedInterpreter& cached_interpreter, u32 data)
+s32 CachedInterpreter::CheckFPU(PowerPC::PowerPCState& ppc_state, const CheckFPUOperands& operands)
 {
-  auto& ppc_state = cached_interpreter.m_ppc_state;
+  const auto& [power_pc, current_pc, downcount] = operands;
   if (!ppc_state.msr.FP)
   {
+    ppc_state.pc = current_pc;
+    ppc_state.downcount -= downcount;
     ppc_state.Exceptions |= EXCEPTION_FPU_UNAVAILABLE;
-    cached_interpreter.m_system.GetPowerPC().CheckExceptions();
-    ppc_state.downcount -= data;
-    return true;
-  }
-  return false;
-}
-
-bool CachedInterpreter::CheckDSI(CachedInterpreter& cached_interpreter, u32 data)
-{
-  auto& ppc_state = cached_interpreter.m_ppc_state;
-  if (ppc_state.Exceptions & EXCEPTION_DSI)
-  {
-    cached_interpreter.m_system.GetPowerPC().CheckExceptions();
-    ppc_state.downcount -= data;
-    return true;
-  }
-  return false;
-}
-
-bool CachedInterpreter::CheckProgramException(CachedInterpreter& cached_interpreter, u32 data)
-{
-  auto& ppc_state = cached_interpreter.m_ppc_state;
-  if (ppc_state.Exceptions & EXCEPTION_PROGRAM)
-  {
-    cached_interpreter.m_system.GetPowerPC().CheckExceptions();
-    ppc_state.downcount -= data;
-    return true;
+    power_pc.CheckExceptions();
+    return 0;
   }
-  return false;
+  return sizeof(AnyCallback) + sizeof(operands);
 }
 
-bool CachedInterpreter::CheckBreakpoint(CachedInterpreter& cached_interpreter, u32 data)
+s32 CachedInterpreter::CheckBreakpoint(PowerPC::PowerPCState& ppc_state,
+                                       const CheckBreakpointOperands& operands)
 {
-  cached_interpreter.m_system.GetPowerPC().CheckBreakPoints();
-  if (cached_interpreter.m_system.GetCPU().GetState() != CPU::State::Running)
+  const auto& [power_pc, cpu_state, current_pc, downcount] = operands;
+  ppc_state.pc = current_pc;
+  if (power_pc.CheckBreakPoints(); *cpu_state != CPU::State::Running)
```

Comment:
Having a special variable for cpu_state when normally you would access it through power_pc feels odd. Is this something you did because it gave a noticeable performance improvement? If so, I suppose we could keep it, but otherwise the cpu_state variable seems redundant.

Reply:
The `CPU::State` value is normally accessed through `CPU::CPUManager::GetState()`/`GetStatePtr()`.  `PowerPC::PowerPCManager` doesn't have a reference to `CPU::CPUManager`, just a private-with-no-getter `Core::System` reference I can't use from `CachedInterpreter` without small modifications to `PowerPC::PowerPCManager` (`Core::System` holds the reference to `CPU::CPUManager`). Regardless, accessing anything through `Core::System` produces bad assembly from what I've observed due to the pointer-to-impl pattern blocking inline-able getters.